includes/discovery/bgp-peers/timos.inc.php

```diff
@@ -83,11 +85,22 @@
             $address = $value['bgpPeerIdentifier'];
 
             if (empty($bgpPeers[$vrfOid][$address])) {
-                $deleted = dbDelete('bgpPeers', 'device_id = ? AND bgpPeerIdentifier = ? AND vrf_id = ?', [$device['device_id'], $address, $vrfId]);
-
+                // if no peer exists with vrf and address matching for this device, then delete
+                if (is_null($vrfId)) {
+                    $deleted = dbDelete('bgpPeers', 'device_id = ? AND bgpPeerIdentifier = ? AND vrf_id is NULL', [$device['device_id'], $address]);
+                } else {
+                    $deleted = dbDelete('bgpPeers', 'device_id = ? AND bgpPeerIdentifier = ? AND vrf_id = ?', [$device['device_id'], $address, $vrfId]);
+                }
                 echo str_repeat('-', $deleted);
                 echo PHP_EOL;
             }
+
+            // deletes any duplicates with same device id, same peer address and same vrfid
+            $duplicatecount = dbFetchCell('SELECT COUNT(*) from `bgpPeers` WHERE device_id = ? AND bgpPeerIdentifier = ? AND vrf_id = ?', [$device['device_id'], $address, $vrfId]);
```

Comment:
Why would there be duplicates?   This troubles me.

Reply:
I will help answer as I worked on the debugging of this problem. We think that discovery of bgppeers adds rows to the database with `vrfID=NULL` . Why these row were being inserted was not found. We couldn't "catch it in the act". But we do now that it doesn't make sense for the Nokia routers to have vrfID to be NULL in the first place.

This lead to the "fix" of deleting these errant rows.

Thus in some ways this row is not a duplicate. It is "wrong data" with the same bdgPeer identifier IP address that is being cleaned up.

Reply:
It should be inserting null not "NULL", the SQL is wrong... One of the downsides of using the ancient db* functions.

To actually insert null, you need to send ['NULL'] as the update parameter. (the string null inside an array)
Or you could just change these to Eloquent queries, but at point, you might as well re-write this as a modern class based module as that would be less work.

Reply:
So the problem is not that the code does not correctly insert NULL, the real problem is that the polling/discovery code for bgppeering should never insert into the database a row where `vrfid` is null. (Also this id is the primary key for the `vrfs` table which the schema is "int(10) unsigned and NULL=no").

Again we don't know why these corrupt data get inserted, but the code change in the PR is to clean this up if found.  I realize it would be ideal to refactor this code, but given the constraints we are under, this bug fix is what we use and wish to contribute to the upstream project.

Reply:
vrf_id should be null in the bgpPeers table when the route is in the default context.

It is not clear that you understand "NULL" and NULL are not the same thing.  Can you please confirm that?
If you look in the database and it says NULL, that means it is a string and not an empty value (null).

It is acceptable to not want to refactor, but I'm having a hard time confirming your bug fix is correct.

Reply:
> It is not clear that you understand "NULL" and NULL are not the same thing. Can you please confirm that?

I am man enough to know when I admit when I am ignorant: I did not understand that you were writing about two different values: a 'null' string and a NULL value. 

However, I believe the values we have in our database that are duplicates are NULL value and not 'null' string. I based this because I ran two different SQL statements, below:

```
# returns nothing
SELECT * FROM bgpPeers WHERE device_id=3697 AND bgpPeerIdentifier="10.10.10.10" AND vrf_id="null";

# returns 4 rows
SELECT * FROM bgpPeers WHERE device_id=3697 AND bgpPeerIdentifier="10.10.10.10" AND vrf_id IS NULL;
```

Your comment above "It should be inserting null not "NULL", the SQL is wrong" now makes more sense.

Reply:
It seems that this line (86) from `timos.inc.php` is the problem? If `$vrfID` is a NULL value then it will not deleting anything. 

```
$deleted = dbDelete('bgpPeers', 'device_id = ? AND bgpPeerIdentifier = ? AND vrf_id = ?', [$device['device_id'], $address, $vrfId]);
```

Reply:
Yes, this seems to be the culprit. 

You can solve this by checking $vrfID and doing 2 different dbDelete depending if vrfID is null or not. Just because `null` in PHP is translated to `''` (empty string) which is not equal to `NULL` in SQL ... 

This check is done in the `vrp.inc.php` bgp implementation, you can have a look at line 142 of includes/discovery/bgp-peers/vrp.inc.php.  Basically, a new variable is created to do the vrf check, and is modified depending if vrfID is null or not.

Reply:
Thanks to you both for the explanations. I'll work with Kate on using the `vrp.inc.php` code as a guide to write the "clean up peers" part of `timos.inc.php` and update this PR.